File: Day1/src/Homework_01/main.js
// 主函数
function main(A) {
    // 加密算法（本质是MD5）
    function J(r) {
        return require("crypto").createHash("md5").update(r).digest("hex");
    }

    function getparam(A) {
        {
            var i = "?";
            for (var e in A)
                i += e,
                    i += "=",
                    i += A[e],
                    i += "&";
            return i = i.substr(0, i.length - 1)
        }
    }

    const Ji = "6RNRDpjjV6wZ2ssPxqeIBeSoV1ITXDdC";

    let i = (new Date).getTime(), e = "";
    // let i = 1679498607223, e = "";
    A.data.timestamp = i.toString();
    "{}" == JSON.stringify(A.data) ? e = `?key=${Ji}` : (e = getparam(A.data), e = `${e}&key=${Ji}`);
    e = (A.url + e).toLowerCase();
    console.log(e);

    return {
        "x-token": J(e),
        "signature": J(J("7Tv7LrinK2bsNAi9TF2uui3ZIcoxK1WT")),
        "params": A.data,
        "e": e
    }
}

console.log(main({
    data: {
        "flag": "top",
        "page": "1",
        "per_page": "9",
    },
    url: "api/product/getProductSearch"
}))

File: Day1/src/Homework_01/main.py
from json import dumps
from execjs import compile
from os.path import join, dirname
import httpx
import asyncio
from urllib3.util import parse_url

endpoint = "https://api.huancang.art/api/product/getProductSearch"

params = {
    "flag": "top",
    "page": "1",
    "per_page": "9",
}


with open(join(dirname(__file__), "./main.js"), mode="r", encoding="UTF-8") as file:
    js_helper = compile(file.read())

encrypted = js_helper.call(
    "main", {"url": parse_url(endpoint).path[1:], "data": params}
)


async def main():
    headers = {
        "authority": "api.huancang.art",
        "accept": "application/json, text/plain, */*",
        "accept-language": "zh-CN,zh;q=0.9,en;q=0.8,en-GB;q=0.7,en-US;q=0.6",
        "cache-control": "no-cache",
        "content-type": "application/json;charset=UTF-8",
        "origin": "https://huancang.art",
        "pragma": "no-cache",
        "referer": "https://huancang.art/",
        "sec-ch-ua": '"Microsoft Edge";v="111", "Not(A:Brand";v="8", "Chromium";v="111"',
        "sec-ch-ua-mobile": "?0",
        "sec-ch-ua-platform": '"Windows"',
        "sec-fetch-dest": "empty",
        "sec-fetch-mode": "cors",
        "sec-fetch-site": "same-site",
        "signature": encrypted["signature"],
        "token": "null",
        "user-agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/111.0.0.0 Safari/537.36 Edg/111.0.1661.44",
        "x-token": encrypted["x-token"],
    }

    async with httpx.AsyncClient() as client:
        response = await client.get(
            endpoint,
            params=encrypted["params"],
            headers=headers,
        )
        print(dumps(response.json(), ensure_ascii=False))


asyncio.run(main())
